pkg/storage/etcd3/v3_helper.go

```diff
@@ -0,0 +1,333 @@
+		}
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	if version, err := h.versioner.ObjectResourceVersion(obj); err == nil && version != 0 {
+		return errors.New("resourceVersion should not be set on objects to be created")
+	}
+	data, err := runtime.Encode(h.codec, obj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	// If rev = 0, it means the key doesn't exist.
+	resp, err := h.client.KV.Txn(ctx).If(
+		clientv3.Compare(clientv3.ModifiedRevision(key), "=", 0),
+	).Then(
+		clientv3.OpPut(key, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	if !resp.Succeeded {
+		return storage.NewKeyExistsError(key, 0)
+	}
+
+	if out != nil {
+		putResp := resp.Responses[0].GetResponsePut()
+		return decode(h.codec, h.versioner, data, out, putResp.Header.Revision)
+
+	}
+	return nil
+}
+
+func (h *v3Helper) Delete(ctx context.Context, key string, out runtime.Object) error {
+	if _, err := conversion.EnforcePtr(out); err != nil {
+		panic("unable to convert output object to pointer")
+	}
+
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	// We need to do get and delete in single transaction in order to
+	// know the value and revision before deleting it.
+	txnResp, err := h.client.KV.Txn(ctx).If().Then(
+		clientv3.OpGet(key),
+		clientv3.OpDelete(key),
+	).Commit()
+	if err != nil {
+		return err
+	}
+	getResp := txnResp.Responses[0].GetResponseRange()
+	if len(getResp.Kvs) == 0 {
+		// To be compatible with original logic of etcdHelper.
+		// In etcd2, delete non-existing key will return error of "key not found"
+		return storage.NewKeyNotFoundError(key, 0)
+	}
+
+	kv := getResp.Kvs[0]
+	return decode(h.codec, h.versioner, kv.Value, out, kv.ModRevision)
+}
+
+func (h *v3Helper) GetToList(ctx context.Context, key string, filter storage.FilterFunc, listObj runtime.Object) error {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+
+	getResp, err := h.client.KV.Get(ctx, key)
+	if err != nil {
+		return err
+	}
+	if len(getResp.Kvs) == 0 {
+		return nil
+	}
+	// Make a list off of the single returned value.
+	values := append([][]byte(nil), getResp.Kvs[0].Value)
+	revs := append([]uint64(nil), uint64(getResp.Kvs[0].ModRevision))
+	if err := decodeList(values, revs, filter, listPtr, h.codec, h.versioner); err != nil {
+		return err
+	}
+	// update version with cluster level revision
+	return h.versioner.UpdateList(listObj, uint64(getResp.Header.Revision))
+}
+
+func (h *v3Helper) List(ctx context.Context, key, resourceVersion string, filter storage.FilterFunc, listObj runtime.Object) error {
+	listPtr, err := meta.GetItemsPtr(listObj)
+	if err != nil {
+		return err
+	}
+	ctx, key = setupRoutine(ctx, key, h.pathPrefix)
+	if key[len(key)-1:] != "/" {
```

Comment:
I guess this is some requirement of the v3 api - can you please document it here?